pkg/compactor/block_upload.go

```diff
@@ -0,0 +1,671 @@
+
+			ooo++
+		}
+
+		/*
+			if ooo > 0 {
+				// TODO: Should this be an error?
+					stats.OutOfOrderSeries++
+					stats.OutOfOrderChunks += ooo
+			}
+		*/
+
+		if err := c.verifyChunks(cr, lset, chnks); err != nil {
+			return err
+		}
+	}
+	if ps.Err() != nil {
+		return errors.Wrap(err, "failed iterating over postings")
+	}
+
+	if outOfOrderLabels > 0 {
+		return errBadRequest{message: fmt.Sprintf("block has %d out of order labels", outOfOrderLabels)}
+	}
+
+	return nil
+}
+
+type errBadRequest struct {
+	message string
+}
+
+func (e errBadRequest) Error() string {
+	return e.message
+}
+
+type errBadGateway struct {
+	message string
+	err     error
+}
+
+func (e errBadGateway) Error() string {
+	return e.message
+}
+
+func (e errBadGateway) Unwrap() error {
+	return e.err
+}
+
+func (c *MultitenantCompactor) verifyChunks(cr *chunks.Reader, lset labels.Labels, chnks []chunks.Meta) error {
+	for _, cm := range chnks {
+		ch, err := cr.Chunk(cm.Ref)
+		if err != nil {
+			return errors.Wrapf(err, "failed to read chunk %d", cm.Ref)
+		}
+
+		samples := 0
+		firstSample := true
+		prevTS := int64(-1)
+
+		it := ch.Iterator(nil)
+		for it.Err() == nil && it.Next() {
+			samples++
+			ts, _ := it.At()
+
+			if firstSample {
+				firstSample = false
+				if ts != cm.MinTime {
+					// TODO: Error?
+					level.Warn(c.logger).Log("ref", cm.Ref, "msg", "timestamp of the first sample doesn't match chunk MinTime",
+						"sampleTimestamp", formatTimestamp(ts), "chunkMinTime", formatTimestamp(cm.MinTime))
+				}
+			} else if ts <= prevTS {
+				// TODO: Error?
+				level.Warn(c.logger).Log("ref", cm.Ref, "msg", "found sample with timestamp not strictly higher than previous timestamp",
+					"previous", formatTimestamp(prevTS), "sampleTimestamp", formatTimestamp(ts))
+			}
+
+			prevTS = ts
+		}
+		if e := it.Err(); e != nil {
+			return errors.Wrapf(err, "failed to failed to iterate over samples of chunk %d", cm.Ref)
+		}
+		if samples == 0 {
+			// TODO: Error?
+			level.Warn(c.logger).Log("ref", cm.Ref, "msg", "no samples found in the chunk")
+		} else if prevTS != cm.MaxTime {
+			// TODO: Error?
+			level.Warn(c.logger).Log("ref", cm.Ref, "msg", "timestamp of the last sample doesn't match chunk MaxTime",
+				"sampleTimestamp", formatTimestamp(prevTS), "chunkMaxTime", formatTimestamp(cm.MaxTime))
+		}
+	}
+
+	return nil
+}
+
+func (c *MultitenantCompactor) sanitizeMeta(tenantID, blockID string, meta *metadata.Meta) error {
+	if meta.Thanos.Labels == nil {
+		meta.Thanos.Labels = map[string]string{}
+	}
+
+	metaULID := meta.ULID
+	if metaULID.String() != blockID {
+		level.Warn(c.logger).Log("msg", "updating meta.json block ID",
+			"old_value", metaULID.String(), "new_value", blockID, "user", tenantID)
+		var err error
+		meta.ULID, err = ulid.Parse(blockID)
+		if err != nil {
+			return errors.Wrapf(err, "couldn't parse block ID %q", blockID)
+		}
+	}
```

Comment:
We can just set it, we don't need to check what the value was before.

Reply:
Done.